network-config-analyzer/IstioTrafficResourcesYamlParser.py

```diff
@@ -0,0 +1,436 @@
+#
+# Copyright 2020- IBM Inc. All rights reserved
+# SPDX-License-Identifier: Apache2.0
+#
+
+from GenericIngressLikeYamlParser import GenericIngressLikeYamlParser
+from IstioTrafficResources import Gateway, VirtualService
+from DimensionsManager import DimensionsManager
+from MinDFA import MinDFA
+from IngressPolicy import IngressPolicy
+from NetworkPolicy import NetworkPolicy
+from Peer import PeerSet
+from functools import reduce
+
+
+class IstioTrafficResourcesYamlParser(GenericIngressLikeYamlParser):
+    """
+    A parser for Istio traffic resources for ingress and egress
+    """
+
+    def __init__(self, peer_container):
+        """
+        :param PeerContainer peer_container: The ingress policy will be evaluated
+        against this set of peers
+        """
+        GenericIngressLikeYamlParser.__init__(self, peer_container)
+        self.namespace = None
+        self.gateways = {}  # a map from a name to a Gateway
+        self.virtual_services = {}  # a map from a name to a VirtualService
+
+    def add_gateway(self, gateway):
+        """
+        Adds a Gateway to the internal parser db
+        (to be later used for locating the gateways referenced by virtual services)
+        :param Gateway gateway: the gateway to add
+        """
+        self.gateways[gateway.full_name()] = gateway
+
+    def get_gateway(self, gtw_full_name):
+        """
+        Returns a Gateway by its full name or None if not found.
+        :param str gtw_full_name: the gateway full name
+        :return Gateway: the found Gateway / None
+        """
+        return self.gateways.get(gtw_full_name)
+
+    def parse_gateway(self, gateway_resource, gateway_file_name):
+        """
+        Parses a gateway resource object and adds the parsed Gateway object to self.gateways
+        :param dict gateway_resource: the gateway object to parse
+        :param str gateway_file_name: the name of the gateway resource file (for reporting errors and warnings)
+        """
+        self.set_file_name(gateway_file_name)  # for error/warning messages
+
+        if gateway_resource.get('kind') != 'Gateway' or 'networking.istio.io' not in gateway_resource.get('apiVersion'):
+            return  # Not an Istio Gateway object
+
+        metadata = gateway_resource.get('metadata')
+        if not metadata:
+            return
+        gtw_name = metadata.get('name')
+        if not gtw_name:
+            return
+        gtw_namespace = self.peer_container.get_namespace(metadata.get('namespace', 'default'))
+        gtw_spec = gateway_resource.get('spec')
+        if not gtw_spec:
+            self.warning(f'Spec is missing or null in Gateway {gtw_name}. Ignoring the gateway')
```

Comment:
It depends. In K8sPolicyYamlParser.py we give warning. In CalicoPolicyYamlParser.py and IstioPolicyYamlParser.py we check it with check_fields_validity and give a syntax_error when does not exit.

Reply:
Also, Adi asked to issue warnings when possible, to not stop the whole run because of a problem in these resources

Reply:
It is extremely weird, but IT IS legal to have a K8s NetworkPolicy without a `spec` field. We do warn about it because it is not best practice.
I'm not sure if Gateway can be defined without a `spec`. You can check this on a live cluster.
Anyway, NCA is supposed to (also) work on the left. e.g., a developer pushes something to git without even knowing if Istio is going to accept it syntactically. NCA can guard from resources with syntax error and block these PRs.

@adisos , any particular reason why we shouldn't report errors for `Gateway` and `VirtualService`?

Reply:
I think it makes sense to report errors for the sections that we parse and support within Gateway/VirtualService (and are considered errors by the spec).